reference-implementation/to-upstream-wpts/transform-streams/examples.js

```diff
@@ -0,0 +1,173 @@
+'use strict';
+
+if (self.importScripts) {
+  self.importScripts('/resources/testharness.js');
+  self.importScripts('../resources/test-utils.js');
+  self.importScripts('../resources/rs-utils.js');
+}
+
+// Generic utility function used by several tests.
+function arrayToReadableStream(array) {
+  return new ReadableStream({
+    start(controller) {
+      for (const chunk of array) {
+        controller.enqueue(chunk);
+      }
+      controller.close();
+    }
+  });
+}
+
+// Example: concatenate an array of readable streams into a single readable stream.
+function concatenateReadables(readables) {
+  const ts = new TransformStream();
+  let promise = Promise.resolve();
+  for (const readable of readables) {
+    promise = promise
+        .then(() => readable.pipeTo(ts.writable, { preventClose: true }),
+              reason => {
+                return Promise.all([
+                  ts.writable.abort(reason),
+                  readable.cancel(reason)]);
+              });
+  }
+  promise.then(() => ts.writable.getWriter().close(),
+               reason => ts.writable.abort(reason))
+      .catch(() => {});
+  return ts.readable;
+}
+
+promise_test(() => {
+  const readable = concatenateReadables([]);
+  return readable.getReader().read().then(({ done }) => {
+    assert_true(done);
+  });
+}, 'concatenateReadables should work with no readables');
+
+promise_test(() => {
+  const readable = concatenateReadables([arrayToReadableStream([0, 1])]);
+  return readableStreamToArray(readable).then(array => assert_array_equals(array, [0, 1], 'array should match'));
+}, 'concatenateReadables should work as a pass-through');
+
+promise_test(() => {
+  const readable = concatenateReadables([arrayToReadableStream([0, 1]), arrayToReadableStream([1, 2])]);
+  return readableStreamToArray(readable).then(array => assert_array_equals(array, [0, 1, 1, 2], 'array should match'));
+}, 'concanateReadables should concatenate readables');
+
+promise_test(t => {
+  const erroredReadable = new ReadableStream({
+    start(controller) {
+      controller.enqueue(2);
+      controller.error();
+    }
+  });
+  const readable = concatenateReadables([arrayToReadableStream([1]), erroredReadable]);
+  const reader = readable.getReader();
+  return reader.read().then(({ value, done }) => {
+    assert_false(done, 'done should be false');
+    assert_equals(value, 1, 'value should be 1');
+    // Error is not passed through due to WritableStreamAbort algorithm step 4.
```

Comment:
Now I'm questioning whether abort not passing through the error is a good idea :(. I think this kind of example was one of the primary reasons I originally designed it to pass through. Oh well...

Reply:
Yes. I am also regretting making `writable.abort()` throw when the stream is errored, as it forces us to stick in empty catch blocks.